File: src/main/scala/jk_5/nailed/server/player/NailedPlayer.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.player

import java.io.{ByteArrayOutputStream, IOException}
import java.util
import java.util.UUID

import com.google.common.collect.ImmutableSet
import io.netty.buffer.Unpooled
import io.netty.util.CharsetUtil
import jk_5.nailed.api.chat.{BaseComponent, ClickEvent, HoverEvent, TextComponent}
import jk_5.nailed.api.map.Map
import jk_5.nailed.api.math.{EulerDirection, Vector3d, Vector3f}
import jk_5.nailed.api.messaging.StandardMessenger
import jk_5.nailed.api.player.Player
import jk_5.nailed.api.plugin.PluginIdentifier
import jk_5.nailed.api.potion.Potion
import jk_5.nailed.api.util.{Checks, Location, TeleportOptions}
import jk_5.nailed.api.world.World
import jk_5.nailed.api.{GameMode, potion}
import jk_5.nailed.server.scoreboard.PlayerScoreboardManager
import jk_5.nailed.server.teleport.Teleporter
import jk_5.nailed.server.{NailedEventFactory, NailedPlatform}
import net.minecraft.entity.SharedMonsterAttributes
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.network.play.server.{S02PacketChat, S3FPacketCustomPayload}
import net.minecraft.network.{NetHandlerPlayServer, Packet}
import net.minecraft.potion.PotionEffect
import net.minecraft.util.DamageSource
import net.minecraft.world.WorldSettings.GameType
import org.apache.logging.log4j.LogManager

import scala.collection.JavaConverters._
import scala.collection.convert.wrapAsScala._
import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
class NailedPlayer(private val uuid: UUID, private var name: String) extends Player {

  var entity: EntityPlayerMP = _
  var world: World = _
  var map: Map = _
  private var displayName: String = this.name
  var netHandler: NetHandlerPlayServer = _
  var isOnline: Boolean = false
  var isAllowedToFly: Boolean = false
  val channels = mutable.HashSet[String]()

  override val getScoreboardManager = new PlayerScoreboardManager(this)

  override def getName = this.name
  override def getDisplayName = this.displayName
  override def getUniqueId = this.uuid
  override def sendMessage(messages: BaseComponent*) = this.netHandler.sendPacket(new S02PacketChat(messages: _*))

  override def heal(amount: Double){
    val newAmount = getHealth + amount
    setHealth(Math.min(getMaxHealth, newAmount))
  }
  override def getMaxHealth = getEntity.getMaxHealth
  override def resetMaxHealth() = setMaxHealth(entity.getMaxHealth)
  override def setMaxHealth(maxHealth: Double){
    Checks.positive(maxHealth, "Max health must be greater than 0")
    entity.getEntityAttribute(SharedMonsterAttributes.maxHealth).setBaseValue(maxHealth)
    if(getHealth > maxHealth){
      setHealth(maxHealth)
    }
  }

  override def getBurnDuration = this.entity.fire
  override def setBurnDuration(ticks: Int) = this.entity.fire = ticks
  override def isBurning = this.entity.fire != 0

  override def setExperience(experience: Double) = entity.experience = experience.toFloat
  override def getExperience = entity.experience.toInt
  override def getLevel = entity.experienceLevel
  override def setLevel(level: Int) = entity.experienceLevel = level

  override def damage(amount: Double): Unit = setHealth(getHealth - amount)
  override def getHealth = Math.min(Math.max(0, getEntity.getHealth), getMaxHealth)
  override def setHealth(health: Double){
    Checks.positiveOrZero(health, "health")
    Checks.smallerThanOrEqual(health, getMaxHealth, "health")
    if(health == 0) entity.onDeath(DamageSource.generic)
    entity.setHealth(health.toFloat)
  }

  override def getPosition = this.getLocation
  override def setPosition(position: Vector3d) = ???
  override def setVectorRotation(rotation: Vector3f) = ???
  override def getVectorRotation = ???
  override def getRotation = ???
  override def setRotation(rotation: EulerDirection) = ???
  override def setVelocity(velocity: Vector3f) = ???
  override def getVelocity = ???
  override def setSaturation(saturation: Double) = ???
  override def setHunger(hunger: Double) = ???
  override def getHunger = ???
  override def getSaturation = ???

  override def teleportTo(world: World){
    Teleporter.teleportPlayer(this, new TeleportOptions(if(world.getConfig != null) Location.builder().copy(world.getConfig.spawnPoint()).setWorld(world).build() else new Location(world, 0, 64, 0)))
  }

  def getEntity = this.entity
  def getWorld = this.world
  def getMap = this.map
  def getLocation = new Location(this.world, entity.posX, entity.posY, entity.posZ, entity.rotationYaw, entity.rotationPitch)

  def sendPacket(packet: Packet) = if(this.netHandler != null) this.netHandler.sendPacket(packet)

  override def getGameMode: GameMode = getEntity.theItemInWorldManager.getGameType match {
    case GameType.SURVIVAL => GameMode.SURVIVAL
    case GameType.CREATIVE => GameMode.CREATIVE
    case GameType.ADVENTURE => GameMode.ADVENTURE
    case GameType.SPECTATOR => GameMode.SPECTATOR
    case e => throw new IllegalStateException("Player has unknown game mode " + e.getName + " " + e.getID)
  }

  override def setGameMode(gm: GameMode): Unit = getEntity.setGameType(gm match {
    case GameMode.SURVIVAL => GameType.SURVIVAL
    case GameMode.CREATIVE => GameType.CREATIVE
    case GameMode.ADVENTURE => GameType.ADVENTURE
    case GameMode.SPECTATOR => GameType.SPECTATOR
  })

  override def setAllowedToFly(allowed: Boolean){
    this.isAllowedToFly = allowed
    this.getEntity.capabilities.allowFlying = allowed
    this.getEntity.capabilities.isFlying = allowed
    this.getEntity.sendPlayerAbilities()
  }

  /*override def getInventorySize: Int = this.getEntity.inventory.getSizeInventory
  override def getInventorySlotContent(slot: Int): ItemStack = ItemStackConverter.toNailed(this.getEntity.inventory.getStackInSlot(slot)) //TODO: maybe save inventories in our system instead the vanilla one
  override def setInventorySlot(slot: Int, stack: ItemStack){
    this.getEntity.inventory.setInventorySlotContents(slot, ItemStackConverter.toVanilla(stack))
    this.getEntity.updateCraftingInventory(this.getEntity.inventoryContainer, this.getEntity.inventoryContainer.getInventory)
  }

  override def addToInventory(stack: ItemStack){
    this.getEntity.inventory.addItemStackToInventory(ItemStackConverter.toVanilla(stack))
    this.getEntity.updateCraftingInventory(this.getEntity.inventoryContainer, this.getEntity.inventoryContainer.getInventory)
  }

  override def iterateInventory(p: ItemStack => Unit){
    for(i <- 0 until this.getInventorySize) p(getInventorySlotContent(i))
  }*/

  override def kick(reason: String){
    this.netHandler.kickPlayerFromServer(reason)
  }

  override def getDescriptionComponent: BaseComponent = {
    val c = new TextComponent(this.getDisplayName)
    c.setHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new TextComponent(this.getUniqueId.toString)))
    c.setClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, "/msg " + this.getName + " "))
    c
  }

  override def addPotionEffect(effect: potion.PotionEffect){
    val vanillaEffect = new PotionEffect(effect.getPotion.getId, effect.getDuration, effect.getLevel - 1, effect.isAmbient, effect.isShowParticles)
    this.entity.addPotionEffect(vanillaEffect)
  }

  override def removePotionEffect(potion: Potion){
    this.entity.removePotionEffect(potion.getId)
  }

  override def clearAllEffects(){
    this.entity.clearActivePotions()
  }

  override def getActiveEffects = {
    val newCollection = ImmutableSet.builder[potion.PotionEffect]()
    for(effect <- this.entity.getActivePotionEffects.asInstanceOf[java.util.Collection[PotionEffect]]){
      val builder = potion.PotionEffect.builder()
      val pot = Potion.byId(effect.getPotionID)
      if(pot != null){
        builder.setPotion(pot)
        builder.setAmbient(effect.getIsAmbient)
        builder.setShowParticles(effect.func_180154_f)
        builder.setLevel(effect.getAmplifier + 1)
        builder.setDuration(effect.getDuration)
        newCollection.add(builder.build())
      }
    }
    newCollection.build()
  }


  override def loadResourcePack(url: String, hash: String){
    this.entity.func_175397_a(url, hash)
  }

  override def sendPluginMessage(source: PluginIdentifier, channel: String, message: Array[Byte]){
    StandardMessenger.validatePluginMessage(NailedPlatform.getMessenger, source, channel, message)
    if(netHandler == null) return

    if(channels.contains(channel)){
      sendPacket(new S3FPacketCustomPayload(channel, Unpooled.copiedBuffer(message)))
    }
  }

  def sendSupportedChannels(){
    if(netHandler == null) return
    val listening = NailedPlatform.getMessenger.getIncomingChannels

    if(!listening.isEmpty){
      val stream = new ByteArrayOutputStream
      for(channel <- listening){
        try{
          stream.write(channel.getBytes(CharsetUtil.UTF_8))
          stream.write(0.toByte)
        }catch{
          case e: IOException => LogManager.getLogger.error("Could not send Plugin Channel REGISTER to " + getName, e)
        }
      }

      sendPacket(new S3FPacketCustomPayload("REGISTER", Unpooled.copiedBuffer(stream.toByteArray)))
    }
  }

  def addChannel(channel: String){
    if(channels.add(channel)){
      NailedEventFactory.firePlayerRegisterChannelEvent(this, channel)
    }
  }

  def removeChannel(channel: String){
    if(channels.remove(channel)){
      NailedEventFactory.firePlayerUnregisterChannelEvent(this, channel)
    }
  }

  def getListeningPluginChannels: util.Set[String] = ImmutableSet.copyOf(channels.asJava: java.lang.Iterable[String])

  override def toString = s"NailedPlayer{uuid=$uuid,name=$name,isOnline=$isOnline,gameMode=$getGameMode,eid=${getEntity.getEntityId}}"
}

File: src/main/java/jk_5/nailed/server/player/NailedPlayer.java
package jk_5.nailed.server.player;

import com.google.common.base.MoreObjects;
import com.google.common.collect.ImmutableSet;
import io.netty.buffer.Unpooled;
import io.netty.util.CharsetUtil;
import jk_5.nailed.api.GameMode;
import jk_5.nailed.api.chat.BaseComponent;
import jk_5.nailed.api.chat.ClickEvent;
import jk_5.nailed.api.chat.HoverEvent;
import jk_5.nailed.api.chat.TextComponent;
import jk_5.nailed.api.chat.serialization.ComponentSerializer;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.map.Team;
import jk_5.nailed.api.math.EulerDirection;
import jk_5.nailed.api.math.Vector3d;
import jk_5.nailed.api.math.Vector3f;
import jk_5.nailed.api.messaging.StandardMessenger;
import jk_5.nailed.api.player.Player;
import jk_5.nailed.api.plugin.PluginIdentifier;
import jk_5.nailed.api.potion.Potion;
import jk_5.nailed.api.potion.PotionEffect;
import jk_5.nailed.api.scoreboard.ScoreboardManager;
import jk_5.nailed.api.util.Checks;
import jk_5.nailed.api.util.Location;
import jk_5.nailed.api.util.TeleportOptions;
import jk_5.nailed.api.util.TitleMessage;
import jk_5.nailed.api.world.World;
import jk_5.nailed.server.NailedEventFactory;
import jk_5.nailed.server.NailedPlatform;
import jk_5.nailed.server.scoreboard.PlayerScoreboardManager;
import jk_5.nailed.server.teleport.Teleporter;
import jk_5.nailed.server.world.NailedWorld;
import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.network.NetHandlerPlayServer;
import net.minecraft.network.Packet;
import net.minecraft.network.play.server.S02PacketChat;
import net.minecraft.network.play.server.S3FPacketCustomPayload;
import net.minecraft.network.play.server.S45PacketTitle;
import net.minecraft.util.DamageSource;
import net.minecraft.util.IChatComponent;
import net.minecraft.world.WorldSettings;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

public class NailedPlayer implements Player {

    private static final Logger logger = LogManager.getLogger();

    private final UUID uuid;
    private final String name;
    private final PlayerScoreboardManager scoreboardManager;
    private final Set<String> channels = new HashSet<String>();

    private EntityPlayerMP entity;
    private String displayName;
    private NetHandlerPlayServer netHandler;
    private boolean isAllowedToFly;
    private boolean isOnline = false;
    private BaseComponent[] subtitle;

    public World world;
    public Map map;

    public NailedPlayer(UUID uuid, String name) {
        this.uuid = uuid;
        this.name = name;
        this.displayName = name;
        this.scoreboardManager = new PlayerScoreboardManager(this);
    }

    @Override
    public ScoreboardManager getScoreboardManager() {
        return this.scoreboardManager;
    }

    @Override
    public String getName() {
        return this.name;
    }

    @Override
    public String getDisplayName() {
        return this.displayName;
    }

    @Override
    public UUID getUniqueId() {
        return this.uuid;
    }

    @Override
    public void sendMessage(BaseComponent... component) {
        if(this.netHandler != null){
            this.netHandler.sendPacket(new S02PacketChat(component));
        }
    }

    @Override
    public void heal(double amount) {
        double newAmount = this.getHealth() + amount;
        this.setHealth(Math.min(this.getMaxHealth(), newAmount));
    }

    @Override
    public double getMaxHealth() {
        return this.entity.getMaxHealth(); //TODO: this does not return the adjusted value (see setMaxHealth)
    }

    @Override
    public void setMaxHealth(double maxHealth) {
        Checks.positive(maxHealth, "Max health must be greater than 0");
        entity.getEntityAttribute(SharedMonsterAttributes.maxHealth).setBaseValue(maxHealth);
        if(this.getHealth() > this.getMaxHealth()){
            setHealth(maxHealth);
        }
    }

    @Override
    public void resetMaxHealth() {
        this.setMaxHealth(this.entity.getMaxHealth());
    }

    @Override
    public boolean isBurning() {
        return this.entity.fire != 0;
    }

    @Override
    public int getBurnDuration() {
        return this.entity.fire;
    }

    @Override
    public void setBurnDuration(int ticks) {
        this.entity.fire = ticks;
    }

    @Override
    public double getExperience() {
        return this.entity.experience;
    }

    @Override
    public int getLevel() {
        return this.entity.experienceLevel;
    }

    @Override
    public void setExperience(double experience) {
        this.entity.experience = (float) experience;
    }

    @Override
    public void setLevel(int level) {
        this.entity.experienceLevel = level;
    }

    @Override
    public void damage(double amount) {
        this.setHealth(this.getHealth() - amount);
    }

    @Override
    public double getHealth() {
        return Math.min(Math.max(0, entity.getHealth()), this.getMaxHealth());
    }

    @Override
    public void setHealth(double health) {
        Checks.positiveOrZero(health, "health");
        Checks.smallerThanOrEqual(health, getMaxHealth(), "health");
        if(health == 0){
            entity.onDeath(DamageSource.generic);
        }
        entity.setHealth((float) health);
    }

    @Nonnull
    @Override
    public Vector3d getPosition() {
        return this.getLocation();
    }

    @Override
    public void setPosition(@Nonnull Vector3d position) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public Vector3f getVectorRotation() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void setVectorRotation(@Nonnull Vector3f rotation) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public EulerDirection getRotation() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void setRotation(@Nonnull EulerDirection rotation) {
        throw new UnsupportedOperationException();
    }

    @Nonnull
    @Override
    public Vector3f getVelocity() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void setVelocity(@Nonnull Vector3f velocity) {
        throw new UnsupportedOperationException();
    }

    @Override
    public double getSaturation() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void setSaturation(double saturation) {
        throw new UnsupportedOperationException();
    }

    @Override
    public double getHunger() {
        return 20 + entity.getFoodStats().getFoodLevel();
    }

    @Override
    public void setHunger(double hunger) {
        entity.getFoodStats().setFoodLevel(20 - (int) hunger);
    }

    @Override
    public void teleportTo(World world) {
        Teleporter.teleportPlayer(this, new TeleportOptions(world.getConfig() != null ? Location.builder().copy(world.getConfig().spawnPoint()).setWorld(world).build() : new Location(world, 0, 64, 0)));
    }

    public EntityPlayerMP getEntity(){
        return this.entity;
    }

    @Nullable
    @Override
    public Map getMap() {
        return this.map;
    }

    @Override
    public World getWorld() {
        return this.world;
    }

    public Location getLocation(){
        return new Location(this.world, entity.posX, entity.posY, entity.posZ, entity.rotationYaw, entity.rotationPitch);
    }

    public void sendPacket(Packet packet){
        if(this.netHandler != null){
            this.netHandler.sendPacket(packet);
        }
    }

    @Override
    public GameMode getGameMode() {
        return GameMode.byId(this.entity.theItemInWorldManager.getGameType().getID());
    }

    @Override
    public void setGameMode(GameMode gameMode) {
        entity.setGameType(WorldSettings.GameType.getByID(gameMode.getId()));
    }

    @Override
    public void setAllowedToFly(boolean allowed) {
        this.isAllowedToFly = allowed;
        this.entity.capabilities.allowFlying = allowed;
        this.entity.capabilities.isFlying = allowed;
        this.entity.sendPlayerAbilities();
    }

    /*override def getInventorySize: Int = this.getEntity.inventory.getSizeInventory
      override def getInventorySlotContent(slot: Int): ItemStack = ItemStackConverter.toNailed(this.getEntity.inventory.getStackInSlot(slot)) //TODO: maybe save inventories in our system instead the vanilla one
      override def setInventorySlot(slot: Int, stack: ItemStack){
        this.getEntity.inventory.setInventorySlotContents(slot, ItemStackConverter.toVanilla(stack))
        this.getEntity.updateCraftingInventory(this.getEntity.inventoryContainer, this.getEntity.inventoryContainer.getInventory)
      }

      override def addToInventory(stack: ItemStack){
        this.getEntity.inventory.addItemStackToInventory(ItemStackConverter.toVanilla(stack))
        this.getEntity.updateCraftingInventory(this.getEntity.inventoryContainer, this.getEntity.inventoryContainer.getInventory)
      }

      override def iterateInventory(p: ItemStack => Unit){
        for(i <- 0 until this.getInventorySize) p(getInventorySlotContent(i))
      }*/

    @Override
    public void kick(@Nonnull String reason) {
        Checks.notNull(reason, "Reason may not be null");
        if(this.netHandler != null){
            this.netHandler.kickPlayerFromServer(reason);
        }
    }

    @Override
    public BaseComponent getDescriptionComponent() {
        TextComponent comp = new TextComponent(this.getDisplayName());
        comp.setHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new TextComponent(this.getUniqueId().toString())));
        comp.setClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, "/msg " + this.getName() + " "));
        return comp;
    }

    @Override
    public void clearAllEffects() {
        this.entity.clearActivePotions();
    }

    @Override
    public void addPotionEffect(@Nonnull PotionEffect effect) {
        this.entity.addPotionEffect(new net.minecraft.potion.PotionEffect(effect.getPotion().getId(), effect.getDuration(), effect.getLevel() - 1, effect.isAmbient(), effect.isShowParticles()));
    }

    @Override
    public void removePotionEffect(@Nonnull Potion potion) {
        this.entity.removePotionEffect(potion.getId());
    }

    @Nonnull
    @Override
    public Collection<PotionEffect> getActiveEffects() {
        ImmutableSet.Builder<PotionEffect> newCollection = ImmutableSet.builder();
        for (net.minecraft.potion.PotionEffect effect : ((Collection<net.minecraft.potion.PotionEffect>) this.entity.getActivePotionEffects())) {
            PotionEffect.Builder builder = PotionEffect.builder();
            Potion pot = Potion.byId(effect.getPotionID());
            if(pot != null){
                builder.setPotion(pot);
                builder.setAmbient(effect.getIsAmbient());
                builder.setShowParticles(effect.getIsShowParticles());
                builder.setLevel(effect.getAmplifier() + 1);
                builder.setDuration(effect.getDuration());
                newCollection.add(builder.build());
            }
        }
        return newCollection.build();
    }

    @Override
    public void loadResourcePack(String url, String hash) {
        this.entity.loadResourcePack(url, hash);
    }

    @Override
    public void sendPluginMessage(PluginIdentifier source, String channel, byte[] message) {
        StandardMessenger.validatePluginMessage(NailedPlatform.instance().getMessenger(), source, channel, message);
        if(netHandler == null){
            return;
        }
        if(channels.contains(channel)){
            this.sendPacket(new S3FPacketCustomPayload(channel, Unpooled.copiedBuffer(message)));
        }
    }

    public void sendSupportedChannels(){
        if(netHandler == null){
            return;
        }
        Set<String> listening = NailedPlatform.instance().getMessenger().getIncomingChannels();
        if(!listening.isEmpty()){
            ByteArrayOutputStream stream = new ByteArrayOutputStream();
            for (String channel : listening) {
                try{
                    stream.write(channel.getBytes(CharsetUtil.UTF_8));
                    stream.write((byte) 0);
                }catch(IOException e){
                    logger.error("Could not send Plugin Channel REGISTER to " + getName(), e);
                }
            }
            sendPacket(new S3FPacketCustomPayload("REGISTER", Unpooled.copiedBuffer(stream.toByteArray())));
        }
    }

    public void addChannel(String channel){
        if(channels.add(channel)){
            NailedEventFactory.firePlayerRegisterChannelEvent(this, channel);
        }
    }

    public void removeChannel(String channel){
        if(channels.remove(channel)){
            NailedEventFactory.firePlayerUnregisterChannelEvent(this, channel);
        }
    }

    @Override
    public Set<String> getListeningPluginChannels(){
        return ImmutableSet.copyOf(channels);
    }

    @Override
    public void displayTitle(@Nonnull TitleMessage title) {
        IChatComponent main = (title.getTitle() != null && title.getTitle().length != 0) ? IChatComponent.Serializer.jsonToComponent(ComponentSerializer.toString(title.getTitle())) : null;
        IChatComponent sub = (title.getSubtitle() != null && title.getSubtitle().length != 0) ? IChatComponent.Serializer.jsonToComponent(ComponentSerializer.toString(title.getSubtitle())) : null;
        sendPacket(new S45PacketTitle(title.getFadeInTime(), title.getDisplayTime(), title.getFadeOutTime()));
        if(main != null){
            this.sendPacket(new S45PacketTitle(S45PacketTitle.Type.TITLE, main));
        }
        if(sub != null){
            this.sendPacket(new S45PacketTitle(S45PacketTitle.Type.SUBTITLE, sub));
        }
    }

    @Override
    public void clearTitle() {
        this.sendPacket(new S45PacketTitle(S45PacketTitle.Type.CLEAR, null));
    }

    @Override
    public void displaySubtitle(BaseComponent... message) {
        this.sendPacket(new S02PacketChat((byte) 2, message));
    }

    @Override
    public void setSubtitle(BaseComponent... message) {
        this.displaySubtitle(message);
        this.subtitle = message;
    }

    @Override
    public void clearSubtitle() {
        this.subtitle = null;
    }

    @Override
    public void clearInventory() {
        entity.inventory.clear();
        //was sendContainerAndContentsToPlayer
        entity.updateCraftingInventory(entity.inventoryContainer, entity.inventoryContainer.getInventory());
    }

    @Override
    public boolean isOnline() {
        return this.isOnline;
    }

    public Location getSpawnPoint(){
        Team team = this.map.getPlayerTeam(this); //TODO: this line throws an NPE when the player logs in for the second time (over EntityPlayerMP)
        if(team == null){
            return ((NailedWorld) world).wrapped().provider.getSpawnPoint();
        }else{
            Location s = team.getSpawnPoint();
            if(s == null){
                return ((NailedWorld) world).wrapped().provider.getSpawnPoint();
            }else{
                return s;
            }
        }
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("uuid", uuid)
                .add("name", name)
                .add("isOnline", isOnline)
                .add("gameMode", this.getGameMode())
                .add("eid", this.entity == null ? -1 : this.entity.getEntityId())
                .toString();
    }
}
